core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
         group <- rawlsGroupQuery if (access.groupName === group.groupName)
         userGroup <- groupUsersQuery if (group.groupName === userGroup.groupName)
         user <- rawlsUserQuery if (user.userSubjectId === userGroup.userSubjectId)
-      } yield (access, user)).map { case (access, user) => (access.accessLevel, user.userEmail) }
+      } yield (access, user)).map { case (access, user) => (access.accessLevel, user.userEmail, user.userSubjectId) }
 
       val accessAndSubGroupEmail = (for {
         access <- workspaceAccessQuery if (access.workspaceId === workspaceContext.workspaceId && access.isRealmAcl === false)
         group <- rawlsGroupQuery if (access.groupName === group.groupName)
         subGroupGroup <- groupSubgroupsQuery if (group.groupName === subGroupGroup.parentGroupName)
         subGroup <- rawlsGroupQuery if (subGroup.groupName === subGroupGroup.childGroupName)
-      } yield (access, subGroup)).map { case (access, subGroup) => (access.accessLevel, subGroup.groupEmail) }
-
-      (accessAndUserEmail union accessAndSubGroupEmail).result.map(_.map { case (access, email) => 
-        (email, WorkspaceAccessLevels.withName(access))
-      })
+      } yield (access, subGroup)).map { case (access, subGroup) => (access.accessLevel, subGroup.groupEmail, subGroup.groupName) }
+
+      /*  The left join here is important. Since we are only going to store share-permission records for users and groups that have been
+          explicitly granted that ability, we want to be able to return all users and groups even if they don't have share permissions.
+          After the join, a null column equates to false, and a non-null column equates to true. This conversion is done with permission.isDefined
+       */
+      val userPermissionQuery = accessAndUserEmail.joinLeft(workspaceUserShareQuery).on(_._3 === _.userSubjectId).map { case (userInfo, permission) => (userInfo._1, userInfo._2, permission.isDefined) }
+      val groupPermissionQuery = accessAndSubGroupEmail.joinLeft(workspaceGroupShareQuery).on(_._3 === _.groupName).map { case (groupInfo, permission) => (groupInfo._1, groupInfo._2, permission.isDefined) }
+
+      userPermissionQuery.result.map(_.map { case (access, email, canShare) =>
+        val accessLevel = WorkspaceAccessLevels.withName(access)
+        // Sharing is implied for owners and project owners, so the permission may not be in the DB. Rectify that here
+        (email, accessLevel, ((accessLevel >= WorkspaceAccessLevels.Owner) || canShare))
+      }).flatMap { userResults =>
+        groupPermissionQuery.result.map(_.map { case (access, email, canShare) =>
+          val accessLevel = WorkspaceAccessLevels.withName(access)
+          // Sharing is implied for owners and project owners, so the permission may not be in the DB. Rectify that here
+          (email, accessLevel, ((accessLevel >= WorkspaceAccessLevels.Owner) || canShare))
```

Comment:
why do we have this logic here instead of in requireSharePermission? This makes it feel like there is an actual row in the db which is maybe changeable. I think I would prefer requireSharePermission have the logic that allows access if the user is >= Owner or canShare